src/session/session_prefetch.c

```diff
@@ -10,55 +10,53 @@
         return (false);
-
     /*
-     * Don't deal with internal pages at the moment - finding the right content to preload based on
-     * internal pages is hard.
+     * Check if pre-fetching is enabled on the session level. We don't perform pre-fetching on
+     * internal threads or internal pages (finding the right content to preload based on internal
+     * pages is hard), so check for that too. We also want to pre-fetch pages that have had at least
+     * one read. The result of this check will be used by cursor logic to determine if pre-fetching
+     * will be performed.
      */
-    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
-        return (false);
-
-    if (session->pf.prefetch_disk_read_count == 1)
-        WT_STAT_CONN_INCR(session, block_prefetch_disk_one);
-
-    /* A single read from disk is common - don't use it to guide pre-fetch behavior. */
-    if (session->pf.prefetch_disk_read_count < 2) {
+    if (!F_ISSET(session, WT_SESSION_PREFETCH) || F_ISSET(session, WT_SESSION_INTERNAL) ||
+      F_ISSET(ref, WT_REF_FLAG_INTERNAL) || session->pf.prefetch_disk_read_count < 2) {
         WT_STAT_CONN_INCR(session, block_prefetch_skipped);
         return (false);
     }
 
-    if (session->pf.prefetch_prev_ref == NULL) {
-        WT_STAT_CONN_INCR(session, block_prefetch_attempts);
-        return (true);
-    }
-
     /*
-     * If the previous pre-fetch was using the same home ref, pre-fetch for approximately the number
-     * of pages that were added to the queue.
+     * We want to avoid the scenario of requesting pre-fetch on one particular ref many times (e.g
+     * when reading along a single page). We can identify this by checking if the previous pre-fetch
+     * was performed using the same home ref.
+     *
+     * In the event that we find this to be true, we perform pre-fetch for approximately the number
+     * of pages that were added to the queue (WT_PREFETCH_QUEUE_PER_TRIGGER). We then want to ensure
+     * that we will not pre-fetch from this ref for a while, and this is done by checking a counter.
+     *
+     * The counter variable prefetch_skipped_with_parent tracks the number of skips we have
+     * performed on a particular ref. If the number of skips surpasses the number of pages that have
+     * been queued for pre-fetch, we are okay to pre-fetch from this ref again. This condition will
+     * evaluate to false and the counter will be reset, effectively marking the ref as available to
+     * pre-fetch from.
      */
     if (session->pf.prefetch_prev_ref->page == ref->home &&
       session->pf.prefetch_skipped_with_parent < WT_PREFETCH_QUEUE_PER_TRIGGER) {
         ++session->pf.prefetch_skipped_with_parent;
         WT_STAT_CONN_INCR(session, block_prefetch_skipped);
         return (false);
     }
+
     session->pf.prefetch_skipped_with_parent = 0;
 
+    if (session->pf.prefetch_disk_read_count == 1)
+        WT_STAT_CONN_INCR(session, block_prefetch_disk_one);
```

Comment:
I don't think this stat can ever get incremented now - we'll bail out before this happens since we now have `|| session->pf.prefetch_disk_read_count < 2` above.

Reply:
You're right - I will move this above `session->pf.prefetch_disk_read_count < 2`